### PR TITLE
feat: Add create possibility pull request for theme builder by GitHub

### PR DESCRIPTION
**Pull request by neretin-trike**
В данном PR реализовал возможность создания пул-реквеста с помощью GitHub API из веб-интрефейса theme-builder'а

:baby_chick: Download canary assets:

[colors--canary.303.4082766000.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.303.4082766000.xml)  
[color_metro_ios-swift--canary.303.4082766000.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_ios-swift--canary.303.4082766000.swift)  
[color_metro_kotlin--canary.303.4082766000.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_kotlin--canary.303.4082766000.kt)  
[color_metro_react-native--canary.303.4082766000.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_react-native--canary.303.4082766000.ts)  
[color_sbermarket_ios-swift--canary.303.4082766000.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.303.4082766000.swift)  
[color_sbermarket_kotlin--canary.303.4082766000.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.303.4082766000.kt)  
[color_sbermarket_react-native--canary.303.4082766000.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.303.4082766000.ts)  
[color_sberprime_ios-swift--canary.303.4082766000.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.303.4082766000.swift)  
[color_sberprime_kotlin--canary.303.4082766000.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.303.4082766000.kt)  
[color_sberprime_react-native--canary.303.4082766000.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.303.4082766000.ts)  
[color_selgros_ios-swift--canary.303.4082766000.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_ios-swift--canary.303.4082766000.swift)  
[color_selgros_kotlin--canary.303.4082766000.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_kotlin--canary.303.4082766000.kt)  
[color_selgros_react-native--canary.303.4082766000.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_react-native--canary.303.4082766000.ts)  
[color_smbusiness_ios-swift--canary.303.4082766000.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_ios-swift--canary.303.4082766000.swift)  
[color_smbusiness_kotlin--canary.303.4082766000.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_kotlin--canary.303.4082766000.kt)  
[color_smbusiness_react-native--canary.303.4082766000.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_react-native--canary.303.4082766000.ts)  
[PlasmaTokensColor--canary.303.4082766000.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.303.4082766000.swift)  
[shadow_sbermarket_react-native--canary.303.4082766000.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/shadow_sbermarket_react-native--canary.303.4082766000.ts)  
[typo_mage_ios-swift--canary.303.4082766000.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_ios-swift--canary.303.4082766000.swift)  
[typo_mage_kotlin--canary.303.4082766000.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.303.4082766000.kt)  
[typo_mage_react-native--canary.303.4082766000.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.303.4082766000.ts)  
[typo_plasma_ios-swift--canary.303.4082766000.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_ios-swift--canary.303.4082766000.swift)  
[typo_plasma_kotlin--canary.303.4082766000.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.303.4082766000.kt)  
[typo_plasma_react-native--canary.303.4082766000.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.303.4082766000.ts)  
[typo_ruler_ios-swift--canary.303.4082766000.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_ios-swift--canary.303.4082766000.swift)  
[typo_ruler_kotlin--canary.303.4082766000.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.303.4082766000.kt)  
[typo_ruler_react-native--canary.303.4082766000.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.303.4082766000.ts)  
[typo_sage_ios-swift--canary.303.4082766000.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_ios-swift--canary.303.4082766000.swift)  
[typo_sage_kotlin--canary.303.4082766000.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.303.4082766000.kt)  
[typo_sage_react-native--canary.303.4082766000.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.303.4082766000.ts)  
[typo_sbermarket_ios-swift--canary.303.4082766000.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_ios-swift--canary.303.4082766000.swift)  
[typo_sbermarket_kotlin--canary.303.4082766000.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.303.4082766000.kt)  
[typo_sbermarket_react-native--canary.303.4082766000.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.303.4082766000.ts)  
[typo_soulmate_ios-swift--canary.303.4082766000.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_ios-swift--canary.303.4082766000.swift)  
[typo_soulmate_kotlin--canary.303.4082766000.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.303.4082766000.kt)  
[typo_soulmate_react-native--canary.303.4082766000.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.303.4082766000.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-b2c@1.134.0-canary.303.4082254687.0
  # or 
  yarn add @salutejs/plasma-b2c@1.134.0-canary.303.4082254687.0
  ```
</details>

### CODE REVIEW

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-303/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-303/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

File: website/plasma-theme-builder/src/api/getFilesSource.ts [hunk 1]
```diff
@@ -16,6 +22,7 @@ export const getFilesSource = async (owner: string, repo: string, path: string,
                 owner,
                 repo,
                 path,
+                ref: branchName,
             });
 
             return result.data;
```
Comment (Yakutoc) [inline website/plasma-theme-builder/src/api/getFilesSource.ts:28]: @neretin-trike Какие нибудь доп проверки нужны? Result может быть nullable?

File: website/plasma-theme-builder/src/api/githubAPIMethods.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
+import { Octokit } from 'octokit';
+
+export interface GitCreateBlobResponse {
+    url?: string;
+    sha?: string | null;
+    path?: string;
+    mode?: '100644' | '100755' | '040000' | '160000' | '120000';
```
Comment (Yakutoc) [inline website/plasma-theme-builder/src/api/githubAPIMethods.ts:7]: [Уточнение]
@neretin-trike А что значат эти цифры? Может комментарий добавить?

File: website/plasma-theme-builder/src/api/githubAPIMethods.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    type?: 'tree' | 'blob' | 'commit';
+}
+
+export const getCurrentSha = async (octokit: Octokit, owner: string, repo: string, branchName: string) => {
```
Comment (Yakutoc) [inline website/plasma-theme-builder/src/api/githubAPIMethods.ts:11]: @neretin-trike Нам нужен здесь try/catch?

File: website/plasma-theme-builder/src/components/PullRequest/PullRequest.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+        const themeName = data?.config.name || 'default';
+        const content = JSON.stringify(data, null, 4) + '\n';
+        const filesTree = getFilesTree(content, themeName);
+
+        const commitType = 'feat';
+        const commitMessage = `Add theme ${themeName}`;
+        const pullRequestHeader = `auto generate ${themeName} theme`;
+
+        const getResult = async () => {
```
Comment (Yakutoc) [inline website/plasma-theme-builder/src/components/PullRequest/PullRequest.tsx:68]: @neretin-trike 

- А что будет если тут будет ошибка? Какое будет поведение?
- try/catch нужен?

File: website/plasma-theme-builder/src/components/PullRequest/PullRequest.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+export const PullRequest = ({ data, token }: PullRequestProps) => {
+    const [step, createPullRequest] = useRunGithubPRProcess({
+        owner: 'salute-developers',
+        repo: 'plasma',
+        branchName: `theme-${nanoid()}`,
+    });
+    const [pullRequestLink, setPullRequestLink] = useState<string | undefined>('');
+
+    useEffect(() => {
+        if (step) {
```
Comment (Yakutoc) [inline website/plasma-theme-builder/src/components/PullRequest/PullRequest.tsx:56]: @neretin-trike А если **step** будет 0? Логика работы не нарушиться?

File: website/plasma-theme-builder/src/api/getFilesSource.ts (hunk 1 above)
Comment (neretin-trike) [reply website/plasma-theme-builder/src/api/getFilesSource.ts:28]: не, если сломается, то сломается с ошибкой и дальше не нужно выполнение кода

File: website/plasma-theme-builder/src/components/PullRequest/PullRequest.tsx (hunk 5 above)
Comment (neretin-trike) [reply website/plasma-theme-builder/src/components/PullRequest/PullRequest.tsx:56]: по умолчанию он как раз 0, что позволяет вызвать getResult один раз, и дальше в этот блок не попадать

File: website/plasma-theme-builder/src/components/PullRequest/PullRequest.tsx (hunk 4 above)
Comment (neretin-trike) [reply website/plasma-theme-builder/src/components/PullRequest/PullRequest.tsx:68]: в случае ошибки он сломается. Не уверен, что try/catch нужен, т.к. дальше смысла выполнять код нету по-идее.

как вариант, можно было бы эти ошибки заворачивать в в нашу страницу с ошибками, но я бы это оставил до момента, когда начну делать редизайн по Гениному макету

File: website/plasma-theme-builder/src/api/githubAPIMethods.ts (hunk 3 above)
Comment (neretin-trike) [reply website/plasma-theme-builder/src/api/githubAPIMethods.ts:11]: написал в этом [коменте](https://github.com/salute-developers/plasma/pull/303#discussion_r1092827454) свои мысли по поводу обработок ошибок

File: website/plasma-theme-builder/src/api/githubAPIMethods.ts (hunk 2 above)
Comment (neretin-trike) [reply website/plasma-theme-builder/src/api/githubAPIMethods.ts:7]: вообще это file mode, т.е. тип, который я ожидаю в репонсе запроса. Вот как в доке это описано:

> one of 100644 for file (blob), 100755 for executable (blob), 040000 for subdirectory (tree), 160000 for submodule (commit), or 120000 for a blob that specifies the path of a symlink

Comment (Yakutoc) [reply website/plasma-theme-builder/src/api/githubAPIMethods.ts:7]: @neretin-trike Может комментарий тогда добавим? Что скажешь?

File: website/plasma-theme-builder/src/components/PullRequest/PullRequest.tsx (hunk 5 above)
Comment (Yakutoc) [reply website/plasma-theme-builder/src/components/PullRequest/PullRequest.tsx:56]: @neretin-trike Ну тогда я правильно все понял. Спасибо за объяснение)

**Review by Yakutoc — APPROVED** [review]

File: website/plasma-theme-builder/src/api/githubAPIMethods.ts (hunk 2 above)
Comment (neretin-trike) [reply website/plasma-theme-builder/src/api/githubAPIMethods.ts:7]: да, подпишу 👌

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-303/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-303/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-303/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-303/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-303/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-303/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-303/temple-storybook/

File: website/plasma-theme-builder/src/api/getEnv.ts [hunk 6]
```diff
@@ -1,4 +1,4 @@
-export const getEnv = () => {
-    const { REACT_APP_APPLICATION_CLIENT_ID = '', REACT_APP_AUTH_SERVER_URL = '' } = process.env;
-    return { REACT_APP_APPLICATION_CLIENT_ID, REACT_APP_AUTH_SERVER_URL };
-};
+export const getEnv = () => ({
+    REACT_APP_APPLICATION_CLIENT_ID: process.env.REACT_APP_APPLICATION_CLIENT_ID || 'a088429392f35252cbc7',
```
[the remark this answers is not in the document]
Comment (Yeti-or) [reply website/plasma-theme-builder/src/api/getEnv.ts:2]: в плане чтобы держать это только локально, я не совсем понимаю на сколько это критичный токен

File: website/plasma-theme-builder/src/api/githubAPIMethods.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    owner: string,
+    repo: string,
+    blobs: GitCreateBlobResponse[],
+    paths: string[],
+    parentTreeSha: string,
+) => {
+    const tree = blobs.map(
+        ({ sha }, index): GitCreateBlobResponse => ({
+            path: paths[index],
+            mode: '100644',
```
Comment (Yeti-or) [inline website/plasma-theme-builder/src/api/githubAPIMethods.ts:61]: тут же есть типизация да?

**Review by Yeti-or — APPROVED** [review]

File: website/plasma-theme-builder/src/api/githubAPIMethods.ts (hunk 7 above)
Comment (neretin-trike) [reply website/plasma-theme-builder/src/api/githubAPIMethods.ts:61]: да, даже коммент поясняющий :)

File: website/plasma-theme-builder/src/api/getEnv.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (neretin-trike) [reply website/plasma-theme-builder/src/api/getEnv.ts:2]: да я бы не стал ради одной переменной тащить целую зависимость. По сути токен относится на прямую к боту, поэтому удобнее её хранить в одном месте рядом с секретом

File: website/plasma-theme-builder/src/components/hooks/useRunGithubPRProcess.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    filesTree: Record<string, string>;
+    token?: string;
+}
+
+const saveMetaData = (octokit: Octokit, owner: string, repo: string) => <T>(fn: (...args: any[]) => Promise<T>) => (
+    ...args: any[]
+) => fn(octokit, owner, repo, ...args);
+
+export const useRunGithubPRProcess = ({ owner, repo, branchName }: RunProcessGithubPR) => {
+    const [step, setStep] = useState<number>(0);
```
[the remark this answers is not in the document]
Comment (neretin-trike) [reply website/plasma-theme-builder/src/components/hooks/useRunGithubPRProcess.ts:32]: хорошая идея, спасиб 👌

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-303/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-303/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-303/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-303/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-303/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-303/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-303/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-303/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-303/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-303/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-303/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-303/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-303/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-303/temple-storybook/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-303/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-303/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-303/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-303/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-303/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-303/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-303/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-303/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-303/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-303/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-303/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-303/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-303/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-303/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-303/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-303/temple-storybook/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-303/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-303/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-303/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-303/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-303/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-303/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-303/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-303/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-303/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-303/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-303/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-303/temple-storybook/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-303/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-303/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-303/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-303/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-303/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-303/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-303/temple-storybook/

**Pull request merged**